WelsonJS.Toolkit/EsentInterop/LibraryHelpers.cs

```diff
@@ -0,0 +1,273 @@
+        [SecurityCritical]
+        private static unsafe IntPtr MyStringToHGlobalUni(string managedString)
+        {
+            if (managedString == null)
+            {
+                return IntPtr.Zero;
+            }
+
+            int charCountWithNull = managedString.Length + 1;
+            int byteCount = charCountWithNull * sizeof(char);
+
+            if (byteCount < managedString.Length)
+            {
+                throw new ArgumentOutOfRangeException("managedString");
+            }
+
+            UIntPtr sizetdwBytes = new UIntPtr((uint)byteCount);
+            IntPtr rawBuffer = Win32.NativeMethods.LocalAlloc(0, sizetdwBytes);
+            if (rawBuffer == IntPtr.Zero)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            fixed (char* sourcePointer = managedString)
+            {
+                byte* destPointer = (byte*)rawBuffer;
+                var unicodeEncoding = new System.Text.UnicodeEncoding();
+                int bytesWritten = unicodeEncoding.GetBytes(sourcePointer, charCountWithNull, destPointer, byteCount);
+            }
+
+            return rawBuffer;
+        }
+#endif // MANAGEDESENT_ON_CORECLR && !MANAGEDESENT_ON_WSA
+
+        /// <summary>Returns a <see cref="T:System.DateTime" /> equivalent to the specified OLE Automation Date.</summary>
+        /// <returns>A <see cref="T:System.DateTime" /> that represents the same date and time as <paramref name="d" />.</returns>
+        /// <param name="d">An OLE Automation Date value. </param>
+        /// <exception cref="T:System.ArgumentException">The date is not a valid OLE Automation Date value. </exception>
+        /// <filterpriority>1</filterpriority>
+        public static DateTime FromOADate(double d)
+        {
+#if MANAGEDESENT_ON_CORECLR
+            return new DateTime(DoubleDateToTicks(d), DateTimeKind.Unspecified);
+#else
+            return DateTime.FromOADate(d);
+#endif
+        }
+
+#if MANAGEDESENT_ON_CORECLR
+        /// <summary>
+        /// Copied from the reflected implementation.
+        /// </summary>
+        /// <param name="value">The date, as a 64bit integer.</param>
+        /// <returns>The date, as a double representation.</returns>
+        internal static double TicksToOADate(long value)
+        {
+            if (value == 0L)
+            {
+                return 0.0;
+            }
+
+            if (value < 864000000000L)
+            {
+                value += 599264352000000000L;
+            }
+
+            if (value < 31241376000000000L)
+            {
+                throw new OverflowException();
+            }
+
+            long num = (value - 599264352000000000L) / 10000L;
+            if (num < 0L)
+            {
+                long num2 = num % 86400000L;
+                if (num2 != 0L)
+                {
+                    num -= (86400000L + num2) * 2L;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

An integer overflow occurs when an arithmetic operation is performed on integer data types without proper checking of maximum value stored in that data type. As C# does not automatically check for or prevent overflow by default, rather it wraps around the result, which may lead to unexpected behaviour or incorrect results. Therefore perform input validation on any numeric input by ensuring that it is within the expected range, when signed integers are required, ensure that the range check includes maximum values. For more information on Integer Overflow refer to https://cwe.mitre.org/data/definitions/190.html.